core/jvm/src/main/scala/fs2/compress.scala

```diff
@@ -75,197 +101,680 @@ object compress {
-        }
+        deflater
+          .end() // If F is narrowed to Applicative then deflater.end() can be bracketed rather than appended in an empty stream.
+        Stream.empty
+      }
+  }
 
-        val body: Stream[F, Byte] = in.chunks.flatMap { c =>
-          processChunk(c)
-          gzos.flush()
-          slurpBytes
-        }
+  private def _gzip_header[F[_]](
+      modificationTime: Option[Instant],
+      deflateLevel: Option[Int],
+      fileName: Option[String],
+      comment: Option[String]
+  ): Stream[F, Byte] = {
+    // See RFC 1952: https://www.ietf.org/rfc/rfc1952.txt
+    val secondsSince197001010000: Long =
+      modificationTime.map(_.getEpochSecond).getOrElse(0)
+    val header = Array[Byte](
+      gzipMagicFirstByte, // ID1: Identification 1
+      gzipMagicSecondByte, // ID2: Identification 2
+      gzipCompressionMethod.DEFLATE, // CM: Compression Method
+      (gzipFlag.FHCRC + // FLG: Header CRC
+        fileName.map(_ => gzipFlag.FNAME).getOrElse(zeroByte) + // FLG: File name
+        comment.map(_ => gzipFlag.FCOMMENT).getOrElse(zeroByte)).toByte, // FLG: Comment
+      (secondsSince197001010000 & 0xff).toByte, // MTIME: Modification Time
+      ((secondsSince197001010000 >> 8) & 0xff).toByte,
+      ((secondsSince197001010000 >> 16) & 0xff).toByte,
+      ((secondsSince197001010000 >> 24) & 0xff).toByte,
+      deflateLevel match { // XFL: Extra flags
+        case Some(Deflater.BEST_COMPRESSION) => gzipExtraFlag.DEFLATE_MAX_COMPRESSION_SLOWEST_ALGO
+        case Some(Deflater.BEST_SPEED)       => gzipExtraFlag.DEFLATE_FASTEST_ALGO
+        case _                               => zeroByte
+      },
+      gzipOperatingSystem.THIS
+    ) // OS: Operating System
+    val crc32 = new CRC32()
+    crc32.update(header)
+    val fileNameEncoded = fileName.map { string =>
+      val bytes = string.replaceAll("\u0000", "_").getBytes(StandardCharsets.ISO_8859_1)
+      crc32.update(bytes)
+      crc32.update(zeroByte)
+      bytes
+    }
+    val commentEncoded = comment.map { string =>
+      val bytes = string.replaceAll("\u0000", " ").getBytes(StandardCharsets.ISO_8859_1)
+      crc32.update(bytes)
+      crc32.update(zeroByte)
+      bytes
+    }
+    val crc32Value = crc32.getValue
+    val crc16 = Array[Byte](
+      (crc32Value & 0xff).toByte,
+      ((crc32Value >> 8) & 0xff).toByte
+    )
+    Stream.chunk(asChunkBytes(header)) ++
+      fileNameEncoded
+        .map(bytes => Stream.chunk(asChunkBytes(bytes)) ++ Stream.emit(zeroByte))
+        .getOrElse(Stream.empty) ++
+      commentEncoded
+        .map(bytes => Stream.chunk(asChunkBytes(bytes)) ++ Stream.emit(zeroByte))
+        .getOrElse(Stream.empty) ++
+      Stream.chunk(asChunkBytes(crc16))
+  }
 
-        val trailer: Stream[F, Byte] = Stream.suspend {
-          gzos.close()
-          slurpBytes
-        }
+  private def _gzip_trailer[F[_]](deflater: Deflater, crc32: CRC32): Stream[F, Byte] =
+    Stream.suspend {
```

Comment:
No need for suspend here as I don't see anything mutable going on

Reply:
Agreed. Stream.suspend removed.